libcontainer/container_linux.go

```diff
@@ -536,6 +537,68 @@ func (c *linuxContainer) NotifyMemoryPressure(level PressureLevel) (<-chan struc
 	return notifyMemoryPressure(c.cgroupManager.GetPaths(), level)
 }
 
+func (c *linuxContainer) checkCriuFeatures(criuOpts *CriuOpts, rpcOpts *criurpc.CriuOpts, criuFeat *criurpc.CriuFeatures) error {
+
+	var t criurpc.CriuReqType
+	t = criurpc.CriuReqType_FEATURE_CHECK
+
+	if err := c.checkCriuVersion("1.8"); err != nil {
+		// Feature checking was introduced with CRIU 1.8.
+		// Ignore the feature check if an older CRIU version is used
+		// and just act as before.
+		// As all automated PR testing is done using CRIU 1.7 this
+		// code will not be tested by automated PR testing.
+		return nil
+	}
+
+	// make sure the features we are looking for are really not from
+	// some previous check
+	c.criuFeat = nil
+
+	req := &criurpc.CriuReq{
+		Type: &t,
+		// Theoretically this should not be necessary but CRIU
+		// segfaults if Opts is empty.
+		// Fixed in CRIU  2.12
+		Opts:     rpcOpts,
+		Features: criuFeat,
+	}
+
+	err := c.criuSwrk(nil, req, criuOpts, false)
+	if err != nil {
+		logrus.Debugf("%s", err)
+		return fmt.Errorf("CRIU feature check failed")
+	}
+
+	logrus.Debugf("Feature check says: %s", c.criuFeat)
+	result := reflect.ValueOf(c.criuFeat)
+	request := reflect.ValueOf(criuFeat)
```

Comment:
Because I do not know what the RPC answer from CRIU will include. It depends on the CRIU version which features it can support and what the answer will include. To not hardcode all possible values in this function I am using reflection to understand which parameters are in the answer and if they match what I wanted to know from CRIU.

It is not really the most readable code, but I do not know how to flexible react on RPC answers without reflect.

The following is an example how it works. Not realistic but good enough to demonstrate how it should work. Instead of asking only for memory_tracking I am also asking for lazy_pages support in a CRIU version which does not know about the lazy RPC feature:

DEBU[0000] Using CRIU 20300 at: /usr/sbin/criu          
DEBU[0000] Using CRIU with following args: [swrk 3]     
DEBU[0000] Using CRIU in FEATURE_CHECK mode             
DEBU[0000] Feature check says: type:FEATURE_CHECK success:true features:<mem_track:true >  
DEBU[0000] Feature check says: mem_track:true           
DEBU[0000] CRIU does not support requested feature LazyPages 
DEBU[0000] CRIU request option LazyPages with value true 
DEBU[0000] CRIU result option LazyPages with value false 
ERRO[0000] container is not destroyed                   
ERRO[0000] CRIU is missing features
                    
CRIU is missing features


So, although I am asking for MemTrack and LazyPages, CRIU answers only with MemTrack (as it does not know anything about LazyPages) and the checkFeature code can handle the answer, even if not all expected fields are filled out.